### PR TITLE
Bug 1880118: Cypress a11y enhancements

### PR DESCRIPTION
**Pull request by dtaylor113**
This PR includes 3 a11y enhancements to our Cypress tests:

**1. Adds a11y checks to Namespaces and Monitoring tests**
**2. Ports Protractors `other-routes` tests to Cypress and does a11y check on the routes:**
```
 Running:  crud/other-routes.spec.ts                                                       (4 of 6)
  Visiting other routes
No BRIDGE_KUBEADMIN_PASSWORD set, skipping login
1 accessibility violation was detected for / page
┌─────────┬─────────────┬───────────┬───────────────────────────────────────┬───────┐
│ (index) │     id      │  impact   │              description              │ nodes │
├─────────┼─────────────┼───────────┼───────────────────────────────────────┼───────┤
│    0    │ 'link-name' │ 'serious' │ 'Ensures links have discernible text' │   5   │
└─────────┴─────────────┴───────────┴───────────────────────────────────────┴───────┘
    ✓ successfully displays view for route: / (10235ms)
    ✓ successfully displays view for route: /k8s/cluster/clusterroles/view (6958ms)
    ✓ successfully displays view for route: /k8s/cluster/nodes (5735ms)
    ✓ successfully displays view for route: /k8s/all-namespaces/events (5852ms)
    ✓ successfully displays view for route: /k8s/all-namespaces/import (5790ms)
    ✓ successfully displays view for route: /api-explorer (7243ms)
    ✓ successfully displays view for route: /api-resource/ns/default/core~v1~Pod (5697ms)
    ✓ successfully displays view for route: /api-resource/ns/default/core~v1~Pod/schema (5656ms)
    ✓ successfully displays view for route: /api-resource/ns/default/core~v1~Pod/instances (5900ms)
    ✓ successfully displays view for route: /api-resource/ns/default/core~v1~Pod/access (5783ms)
    ✓ successfully displays view for route: /k8s/cluster/user.openshift.io~v1~User (5659ms)
    ✓ successfully displays view for route: /k8s/ns/openshift-machine-api/machine.openshift.io~v1beta1~Machine (5756ms)
    ✓ successfully displays view for route: /k8s/ns/openshift-machine-api/machine.openshift.io~v1beta1~MachineSet (5680ms)
    ✓ successfully displays view for route: /k8s/ns/openshift-machine-api/autoscaling.openshift.io~v1beta1~MachineAutoscaler (5824ms)
    ✓ successfully displays view for route: /k8s/ns/openshift-machine-api/machine.openshift.io~v1beta1~MachineHealthCheck (5730ms)
    ✓ successfully displays view for route: /k8s/cluster/machineconfiguration.openshift.io~v1~MachineConfig (5734ms)
    ✓ successfully displays view for route: /k8s/cluster/machineconfiguration.openshift.io~v1~MachineConfigPool (5704ms)
    ✓ successfully displays view for route: /k8s/all-namespaces/monitoring.coreos.com~v1~Alertmanager (5683ms)
    ✓ successfully displays view for route: /k8s/ns/openshift-monitoring/monitoring.coreos.com~v1~Alertmanager/main (5641ms)
    ✓ successfully displays view for route: /settings/cluster (5679ms)
    ✓ successfully displays view for route: /monitoring/query-browser (5671ms)
No BRIDGE_KUBEADMIN_PASSWORD set, skipping logout
    ✓ successfully displays view for route: /search/all-namespaces?kind=config.openshift.io~v1~Console (6087ms)
```
**3. Outputs a11y summary:**
```
A11y Test Results:
┌──────────────────────┬────────┐
│       (index)        │ Values │
├──────────────────────┼────────┤
│   numberA11yChecks   │  '82'  │
│ numberA11yViolations │  '4'   │
└──────────────────────┴────────┘
```
JIRA Story: https://issues.redhat.com/browse/CONSOLE-2359

### CODE REVIEW

File: frontend/packages/integration-tests-cypress/plugins/index.js [hunk 1]
```diff
@@ -33,6 +34,12 @@ module.exports = (on, config) => {
       console.table(data);
       return null;
     },
+    readFileMaybe(filename) {
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/plugins/index.js:37]:
```suggestion
    readFileIfExists(filename) {
```

File: frontend/packages/integration-tests-cypress/support/a11y.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+Cypress.Commands.add('logA11yViolations', (violations: Result[], target: string) => {
+  // pluck specific keys to keep the table readable
+  const violationData = violations.map(({ id, impact, description, nodes }) => ({
+    id,
+    impact,
+    description,
+    nodes: nodes.length,
+  }));
+  window.Cypress.numberA11yViolations += violations.length;
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/support/a11y.ts:22]: Should we be modifying the `Cypress` object? I'd think we'd want to avoid mutating that since it's not our API and store this somewhere else.

File: frontend/public/declarations.d.ts [hunk 3]
```diff
@@ -45,7 +45,10 @@ declare interface Window {
   __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: Function;
   store?: {}; // Redux store
   pluginStore?: {}; // Console plugin store
-  Cypress?: {};
+  Cypress?: {
+    numberA11yViolations: number;
+    numberA11yChecks: number;
```
Comment (spadgett) [inline frontend/public/declarations.d.ts:50]: If we're only accessing these in the Cypress tests, it would be better to avoid adding the types here since this `window` definition applies to console itself.

File: frontend/packages/integration-tests-cypress/support/index.ts [hunk 4]
```diff
@@ -3,57 +3,40 @@ import './project';
...
+      } catch (e) {
+        cy.task(
+          'logError',
+          `couldn't parse cypress-a11y-results.json.  ${JSON.stringify(e, null, ' ')}`,
+        );
+        window.Cypress.numberA11yViolations = 0;
+        window.Cypress.numberA11yChecks = 0;
+      }
+      window.Cypress.numberA11yViolations = Number(a11yResult.numberA11yViolations);
+      window.Cypress.numberA11yChecks = Number(a11yResult.numberA11yChecks);
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/support/index.ts:27]: If you move this inside the `try` block, you can use `const` on `allyResult`.

File: frontend/packages/integration-tests-cypress/support/index.ts [hunk 5]
```diff
@@ -3,57 +3,40 @@ import './project';
...
+before(() => {
+  cy.task('readFileMaybe', 'cypress-a11y-results.json').then((a11yResultsOrNull) => {
+    if (a11yResultsOrNull !== null) {
+      let a11yResult;
+      try {
+        a11yResult = JSON.parse(a11yResultsOrNull);
+      } catch (e) {
+        cy.task(
+          'logError',
+          `couldn't parse cypress-a11y-results.json.  ${JSON.stringify(e, null, ' ')}`,
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/support/index.ts:21]: I don't think we need the full stack trace here, only the error message?

```suggestion
          `couldn't parse cypress-a11y-results.json.  ${e}`,
```

File: frontend/packages/integration-tests-cypress/tests/crud/k8-openshift-cruds.spec.ts [hunk 6]
```diff
@@ -84,7 +84,10 @@ describe('Kubernetes resource CRUD operations', () => {
         // sidebar needs to be fully loaded, else it sometimes overlays the Create button
         cy.byTestID('resource-sidebar').should('exist');
         yamlEditor.isLoaded();
-        cy.testA11y(`YAML Editor for ${kind}: ${name}`);
+        // YAML Editor same for all resources, so only a11y test once
+        if (kind === 'Pod') {
+          cy.testA11y(`YAML Editor for ${kind}: ${name}`);
+        }
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/crud/k8-openshift-cruds.spec.ts:90]: Is there a reason we shouldn't test for every resource? I could see us adding special handling for certain resource. We're already talking about adding a YAML/Form switch for some resources.

File: frontend/packages/integration-tests-cypress/tests/crud/k8-openshift-cruds.spec.ts [hunk 7]
```diff
@@ -132,8 +135,10 @@ describe('Kubernetes resource CRUD operations', () => {
         // filter should have 3 chip groups: resource, label, and name
         listPage.filter.numberOfActiveFiltersShouldBe(3);
         listPage.rows.shouldExist(name);
-        cy.testA11y(`Search page for ${kind}: ${name}`);
-
+        // Search page same for all resources, so only a11y test once
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/crud/k8-openshift-cruds.spec.ts:138]: This isn't true since the lists are different for each resource. I'd go ahead and test each one unless there is a good reason not to.

File: frontend/packages/integration-tests-cypress/tests/crud/namespace-crud.spec.ts [hunk 8]
```diff
@@ -25,11 +25,13 @@ describe('Namespace', () => {
     listPage.rows.shouldNotExist(newName);
     listPage.filter.byName(testName);
     listPage.rows.shouldExist(testName); // created via cy.createProject(testName) above
+    cy.testA11y(`Namespace List page`);
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/crud/namespace-crud.spec.ts:28]: Unnecessary backticks. (I thought we had an eslint rule for this.)

```suggestion
    cy.testA11y('Namespace List page');
```

File: frontend/packages/integration-tests-cypress/tests/crud/namespace-crud.spec.ts [hunk 9]
```diff
@@ -25,11 +25,13 @@ describe('Namespace', () => {
...
 
     cy.log('creates the Namespace');
     listPage.clickCreateYAMLbutton();
     modal.shouldBeOpened();
     cy.byTestID('input-name').type(newName);
+    cy.testA11y(`Create Namespace modal`);
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/crud/namespace-crud.spec.ts:34]:
```suggestion
    cy.testA11y('Create Namespace modal');
```

File: frontend/packages/integration-tests-cypress/tests/crud/namespace-crud.spec.ts [hunk 10]
```diff
@@ -41,6 +43,7 @@ describe('Namespace', () => {
     listPage.rows.clickKebabAction(newName, 'Delete Namespace');
     modal.shouldBeOpened();
     cy.byTestID('project-name-input').type(newName);
+    cy.testA11y(`Delete Namespace modal`);
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/crud/namespace-crud.spec.ts:46]:
```suggestion
    cy.testA11y('Delete Namespace modal');
```

File: frontend/packages/integration-tests-cypress/tests/monitoring/monitoring.spec.ts [hunk 11]
```diff
@@ -45,6 +45,7 @@ describe('Monitoring: Alerts', () => {
     cy.byLegacyTestID('resource-title').should('have.text', 'Alerting');
     listPage.projectDropdownShouldNotExist();
     listPage.rows.shouldBeLoaded();
+    cy.testA11y(`Monitor Alerting list page`);
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/monitoring/monitoring.spec.ts:48]: Unnecessary backticks throughout this file.

File: frontend/packages/integration-tests-cypress/tests/crud/k8-openshift-cruds.spec.ts (hunk 6 above)
Comment (dtaylor113) [reply frontend/packages/integration-tests-cypress/tests/crud/k8-openshift-cruds.spec.ts:90]: I thought the YAML Editor html elements would be the same for every resource, only the inner YAML text changes, so no need to rerun a11y checks for every resource, just one would do, right?

File: frontend/packages/integration-tests-cypress/tests/crud/k8-openshift-cruds.spec.ts (hunk 7 above)
Comment (dtaylor113) [reply frontend/packages/integration-tests-cypress/tests/crud/k8-openshift-cruds.spec.ts:138]: ok, thought it was the same for every resource.

File: frontend/packages/integration-tests-cypress/support/a11y.ts (hunk 2 above)
Comment (dtaylor113) [reply frontend/packages/integration-tests-cypress/support/a11y.ts:22]: Ok, how about `window.a11y.numberViolations` ?

**Comment by dtaylor113** [comment]
Hi @spadgett, I believe I have addressed all of your comments.  PTAL  -thanks

File: frontend/packages/integration-tests-cypress/support/index.ts [hunk 12]
```diff
@@ -1,59 +1,37 @@
+/* eslint-disable import/no-duplicates */
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/support/index.ts:1]: Why is this needed?

File: frontend/packages/integration-tests-cypress/tests/crud/k8-openshift-cruds.spec.ts (hunk 6 above)
Comment (spadgett) [reply frontend/packages/integration-tests-cypress/tests/crud/k8-openshift-cruds.spec.ts:90]: > I thought the YAML Editor html elements would be the same for every resource, only the inner YAML text changes, so no need to rerun a11y checks for every resource, just one would do, right?

I wouldn't assume that will always be true, and there's no virtually no overhead to checking a11y for each resource. And it makes the code simpler as well. I don't see a reason to restrict it to just pods.

**Comment by dtaylor113** [comment]
> spadgett 2 minutes ago Member
I thought the YAML Editor html elements would be the same for every resource, only the inner YAML text changes, so no need to rerun a11y checks for every resource, just one would do, right?

>> I wouldn't assume that will always be true, and there's no virtually no overhead to checking a11y for each resource. And it makes the code simpler as well. I don't see a reason to restrict it to just pods.

Ok, just didn't want to inflate a11y total number of checks run, that's an extra 27 checks which might be the same page/elements.

File: frontend/packages/integration-tests-cypress/support/index.ts (hunk 12 above)
Comment (dtaylor113) [reply frontend/packages/integration-tests-cypress/support/index.ts]: It is not :-) removed.

**Review by spadgett — APPROVED** [review]
/lgtm

Thanks @dtaylor113 👍

File: frontend/packages/integration-tests-cypress/tests/crud/other-routes.spec.ts [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+          '/settings/cluster',
+          '/monitoring/query-browser',
+          // Test loading search page for a kind with no static model.
+          '/search/all-namespaces?kind=config.openshift.io~v1~Console',
+        ]
+      : []),
+  ];
+  otherRoutes.forEach((route) => {
+    it(`successfully displays view for route: ${route}`, () => {
+      cy.visit(`${route}`);
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/crud/other-routes.spec.ts:47]: nit:

```suggestion
      cy.visit(route);
```

Comment (dtaylor113) [reply frontend/packages/integration-tests-cypress/tests/crud/other-routes.spec.ts]: Hi @spadgett, fixed, pls re-add 'lgtm' label -thanks

**Review by spadgett — APPROVED** [review]
/lgtm

**Comment by dtaylor113** [comment]
Hi @spadgett, had to resolve merge conflict, pls re-add 'lgtm' label -thanks

**Review by spadgett — APPROVED** [review]
/lgtm
